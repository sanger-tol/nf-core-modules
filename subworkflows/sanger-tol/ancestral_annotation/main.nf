//
// Create ancestral linkage plots against input assembly
//
include { SAMTOOLS_FAIDX        } from '../../../modules/nf-core/samtools/faidx/main'
include { ANCESTRAL_EXTRACT     } from '../../../modules/sanger-tol/ancestral/extract'
include { ANCESTRAL_PLOT        } from '../../../modules/sanger-tol/ancestral/plot'


workflow ANCESTRAL_ANNOTATION {
    take:
    reference            // Channel: [ meta, reference ]
    ancestral_table      // Channel: [ meta, ancestral_table ]
    busco_full_table     // Channel: [ meta, busco_dir ]

    main:
    ch_versions                     = Channel.empty()


    //
    // MODULE: EXTRACTS ANCESTRALLY LINKED BUSCO GENES FROM FULL TABLE
    //
    ANCESTRAL_EXTRACT(
        busco_full_table,
        ancestral_table
    )
    ch_versions                     = ch_versions.mix(ANCESTRAL_EXTRACT.out.versions)


    //
    // MODULE: INDEX THE INPUT ASSEMBLY
    //
    SAMTOOLS_FAIDX(
        reference,
        [[],[]],
        false
    )
    ch_versions                     = ch_versions.mix( SAMTOOLS_FAIDX.out.versions )


    //
    // MODULE: PLOTS THE ANCESTRAL BUSCO GENES
    //
    ANCESTRAL_PLOT (
        ANCESTRAL_EXTRACT.out.comp_location,
        SAMTOOLS_FAIDX.out.fai
    )
    ch_versions                     = ch_versions.mix(ANCESTRAL_PLOT.out.versions)


    emit:
    ancestral_png_plot              = ANCESTRAL_PLOT.out.png_plot           // channel: [   [id], file  ]
    ancestral_pdf_plot              = ANCESTRAL_PLOT.out.pdf_plot           // channel: [   [id], file  ]
    ancestral_complete_location     = ANCESTRAL_EXTRACT.out.comp_location   // channel: [   [id], file  ]
    ancestral_duplicate_location    = ANCESTRAL_EXTRACT.out.dup_location    // channel: [   [id], file  ]
    ancestral_summary               = ANCESTRAL_EXTRACT.out.summary         // channel: [   [id], file  ]
    versions                        = ch_versions                           // channel: [ versions.yml  ]

}
