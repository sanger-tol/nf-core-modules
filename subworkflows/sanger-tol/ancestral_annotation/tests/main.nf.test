nextflow_workflow {

    name "Test Subworkflow ANCESTRAL_ANNOTATION"
    script "../main.nf"
    config "./nextflow.config"
    workflow "ANCESTRAL_ANNOTATION"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/ancestral_annotation"
    tag "samtools/faidx"
    tag "ancestral/extract"
    tag "ancestral/plot"
    tag "subworkflows/../../modules/nf-core/samtools/faidx"
	tag "modules/nf-core/gunzip"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "samtools/faidx",
                "gunzip"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")

        run("GUNZIP"){
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id: "Ceramica_pisi" ],
                    file(params.modules_testdata_base_path + 'Ceramica_pisi/assembly/release/ilCerPisi1.1/insdc/GCA_963859965.1.fasta.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("Ceramica_pisi genome [fasta]") {

        when {

            params {
                ancestral_plot_args     = "-m True -d True"
            }

            workflow {
                """
                input[0] = GUNZIP.out.gunzip

                input[1] = Channel.of(
                    [
                        [ id:'lepidoptera' ], // meta map
                        file(params.modules_testdata_base_path + 'resources/genomenote/ancestral_data/Merian_elements_full_table.tsv', checkIfExists: true)
                    ]
                )

                input[2] = Channel.of(
                    [
                        [ id:'test-full-table' ], // meta map
                        file(params.modules_testdata_base_path + 'resources/genomenote/ancestral_data/GCA_963859965.1_busco_full_table.tsv', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.ancestral_png_plot.get(0).get(1)).getName(),
                    file(workflow.out.ancestral_png_plot.get(0).get(1)).size(),
                    file(workflow.out.ancestral_pdf_plot.get(0).get(1)).getName(),
                    file(workflow.out.ancestral_pdf_plot.get(0).get(1)).size(),
                    file(workflow.out.ancestral_complete_location.get(0).get(1)),
                    file(workflow.out.ancestral_duplicate_location.get(0).get(1)),
                    file(workflow.out.ancestral_summary.get(0).get(1)),
                    workflow.out.versions
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("Cloeon_dipterum genome - stub") {
        options "-stub"
        when {

            params {
                ancestral_plot_args     = "-m True -d True"
            }

            workflow {
                """
                input[0] = GUNZIP.out.gunzip

                input[1] = Channel.of(
                    [
                        [ id:'lepidoptera' ], // meta map
                        file(params.modules_testdata_base_path + 'resources/genomenote/ancestral_data/Merian_elements_full_table.tsv', checkIfExists: true)
                    ]
                )

                input[2] = Channel.of(
                    [
                        [ id:'test-full-table' ], // meta map
                        file(params.modules_testdata_base_path + 'resources/genomenote/ancestral_data/GCA_963859965.1_busco_full_table.tsv', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.ancestral_png_plot.get(0).get(1)).getName(),
                    file(workflow.out.ancestral_pdf_plot.get(0).get(1)).getName(),
                    file(workflow.out.ancestral_complete_location.get(0).get(1)),
                    file(workflow.out.ancestral_duplicate_location.get(0).get(1)),
                    file(workflow.out.ancestral_summary.get(0).get(1)),
                    workflow.out.versions
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }
}
