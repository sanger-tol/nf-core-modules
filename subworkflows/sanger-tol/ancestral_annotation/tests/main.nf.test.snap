{
    "Cloeon_dipterum genome - stub": {
        "content": [
            "test-full-table_buscopainter_complete_location.tsv_buscopainter.png",
            "test-full-table_buscopainter_complete_location.tsv_buscopainter.pdf",
            "test-full-table_buscopainter_complete_location.tsv:md5,d41d8cd98f00b204e9800998ecf8427e",
            "test-full-table_buscopainter_duplicated_location.tsv:md5,d41d8cd98f00b204e9800998ecf8427e",
            "test-full-table_summary.tsv:md5,d41d8cd98f00b204e9800998ecf8427e",
            [
                "versions.yml:md5,20310e7fc172ac22d25be609af42106c",
                "versions.yml:md5,d0ef9468db0e5edbca7a780f944e3511",
                "versions.yml:md5,d5b755683b719e0d350013bdbfc1214b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-10-08T16:35:43.743911115"
    },
    "Ceramica_pisi genome [fasta]": {
        "content": [
            "buscopainter_complete_location.tsv_buscopainter.png",
            "buscopainter_complete_location.tsv_buscopainter.pdf",
            "buscopainter_complete_location.tsv:md5,92a2d60e7a66edeb5070e3d982a53193",
            "buscopainter_duplicated_location.tsv:md5,23301496a69958b2b64d30df982ca2df",
            "buscopainter_summary.tsv:md5,16931cc78cc8506863b360e149d81111",
            [
                "versions.yml:md5,20310e7fc172ac22d25be609af42106c",
                "versions.yml:md5,44f213ae69f976c93a19fd80e0520cae",
                "versions.yml:md5,d0ef9468db0e5edbca7a780f944e3511"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-10-08T16:54:30.303804678"
    }
}